File: 1_funcs/funcional1.kt
fun suma(a: Int, b: Int): Int{
    return a + b
}

fun multiplica (a: Int, b: Int): Int{
    return a * b
}


fun main() {

    val operations = object {
        val _suma = ::suma
        val _multiplica = ::multiplica
    }

    println("sua suma:: ${operations._suma(2, 3)}")
    println("sua multiplica:: ${operations._multiplica(2, 3)}")

    // println("sua multiplica:: ${operations.multiplica(2,3)}")

}

//12 nov
//blocked, copy from jS example
//https://stackoverflow.com/questions/64798614/kotlin-functions-inside-a-const


//
//higher order functions
//funcs that operate on other funcs
// take other funcs as args
// returns other funcs

val numbers = setOf(1, 1, 2, 3, 5, 8)

fun transformFunction (x: Int): Int {
    return x + 2
}

println("transformFunction:: ${numbers.map{transformFunction}}")

File: 1_funcs/funcional1.js

function suma (a, b){
    return a + b
}

console.log("sua suma::", suma(2,3))

function multiplica (a, b){
    return a * b
}

console.log("sua multiplik::", multiplica(2,3))

//first class functions
// 1. pass around
// 2. can be assigned to variables
// 3. stored in more complex data structures

const operations = {
    suma,
    multiplica
}

console.log ("sum of first class::", operations.suma(2,3))
console.log ("mult of first class::", operations.multiplica(2,3))

//
//higher order functions
//funcs that operate on other funcs
// take other funcs as args
// returns other funcs

const numbers = [1, 1, 2, 3, 5, 8]
const transformFunction = x => x + 2

console.log ("originale    :: ", numbers)
console.log ("transformatio:: ", numbers.map(transformFunction))

// return func from other func

const makeGreeter = greeting => name => `${greeting}, ${name}!`

const greet = makeGreeter('haho')
console.log(greet('mako'))





//notes
//practice: map, reduce, filter, compose, forEach, etc